File: src/main/java/kr/devport/api/security/CustomUserDetails.java
package kr.devport.api.security;

import kr.devport.api.domain.entity.User;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;

@AllArgsConstructor
@Getter
public class CustomUserDetails implements UserDetails, OAuth2User {

    private Long id;
    private String email;
    private String name;
    private Collection<? extends GrantedAuthority> authorities;
    private Map<String, Object> attributes;

    public static CustomUserDetails create(User user) {
        Collection<GrantedAuthority> authorities = Collections.singletonList(
            new SimpleGrantedAuthority("ROLE_" + user.getRole().name())
        );

        return new CustomUserDetails(
            user.getId(),
            user.getEmail(),
            user.getName(),
            authorities,
            null
        );
    }

    public static CustomUserDetails create(User user, Map<String, Object> attributes) {
        CustomUserDetails userDetails = create(user);
        return new CustomUserDetails(
            userDetails.getId(),
            userDetails.getEmail(),
            userDetails.getName(),
            userDetails.getAuthorities(),
            attributes
        );
    }

    @Override
    public String getPassword() {
        return null; // OAuth2 사용자는 별도 비밀번호가 없다.
    }

    @Override
    public String getUsername() {
        return email;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    @Override
    public Map<String, Object> getAttributes() {
        return attributes;
    }
}

File: support/security/src/main/kotlin/kr/devport/api/domain/common/security/CustomUserDetails.kt
package kr.devport.api.domain.common.security

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.core.user.OAuth2User

/**
 * Authenticated principal. A plain holder with no domain dependency — the auth module builds it
 * from its `User` entity via a factory, keeping this shared type free of domain coupling.
 */
class CustomUserDetails(
    val id: Long,
    val email: String?,
    private val usernameValue: String?,
    private val passwordValue: String?,
    private val nameValue: String?,
    private val authoritiesValue: Collection<out GrantedAuthority>?,
    private val attributesValue: Map<String, Any>?,
) : UserDetails,
    OAuth2User {
    fun withAttributes(attributes: Map<String, Any>?): CustomUserDetails =
        CustomUserDetails(id, email, usernameValue, passwordValue, nameValue, authoritiesValue, attributes)

    override fun getUsername(): String = usernameValue ?: ""

    override fun getPassword(): String? = passwordValue

    override fun getName(): String = nameValue ?: ""

    override fun getAuthorities(): Collection<out GrantedAuthority> = authoritiesValue ?: emptyList()

    override fun getAttributes(): Map<String, Any> = attributesValue ?: emptyMap()

    override fun isAccountNonExpired(): Boolean = true

    override fun isAccountNonLocked(): Boolean = true

    override fun isCredentialsNonExpired(): Boolean = true

    override fun isEnabled(): Boolean = true
}
